### DISCUSSION TITLE
Using S3 Storage, but file URL still from application URL

### DISCUSSION THREAD

**Discussion by awd3321** [pocketbase/pocketbase#4447] [Q&A]
I already setup all required S3 storage and connection success, when i upload a files, it appears in my collections and my s3 storage.
The problem is the file urls is from the application URL and not from the s3 generated url.

Im using ```pb.files.getUrl(record, firstFilename, {'thumb': '100x250'});``` to get file url as mentioned in the docs, using js sdk.
How can i get the url and download the file generated url from s3 storage?
Normally by adding ```?download=1``` right?

**Comment by ganigeorgiev** [comment]
_See https://github.com/pocketbase/pocketbase/discussions/4046, https://github.com/pocketbase/pocketbase/discussions/4332, https://github.com/pocketbase/pocketbase/discussions/2191 and probably others._

> The problem is the file urls is from the application URL and not from the s3 generated url.

This should be expected. PocketBase files are served from the `/api/files/*` endpoint no matter whether you use the local filesystem or S3 (_we can't expose directly the S3 url because the file could be private and all S3 operations, including file download, are performed through the S3 APIs using the configured credentials on the PocketBase server-side_).

You are free to use directly the S3 endpoint with the `https://yourPublicS3Url/:collectionId/:recordId/:filename` format if you want to but note that your bucket will have to be always public and the thumbs may not work because they are generated lazily by the PocketBase server when you request them for the first time from `/api/files/*`.

**Reply by BanDroid** [reply]
hi, i just wanna ask, if i switch from s3 as storage to my local again, can my app still access the files that are in s3 through pocketbase?

**Reply by ganigeorgiev** [reply]
@BanDroid No, PocketBase no longer operates with S3 once the toggle is disabled. You need to download the existing files from S3 and place them manually inside `pb_data/storage/` directory (_the files organization is the same as the uploaded object prefix where each `/` separated segment is a sub directory_).

External tools like [rclone](https://github.com/rclone/rclone) or [s5cmd](https://github.com/peak/s5cmd) can help you automate the process.

**Reply by BanDroid** [reply]
could you make a special option in file input to toggle between s3 and local when creating new record? so each record could either choose to store in s3 or local differently, make the storage more efficient and easy to switch sides

**Reply by ganigeorgiev** [reply]
No, it will complicate unnecessary too many things.
